File: web-app/tests/FxRepository.test.ts
import { describe, it, expect, vi } from 'vitest';
import { FxRepository } from '../src/repositories/FxRepository';

describe('FxRepository', () => {
  it('caches rates from service', async () => {
    const getRate = vi.fn().mockResolvedValue(1.1);
    const repo = new FxRepository({ getRate } as any);
    const first = await repo.rate('USD', 'EUR');
    const second = await repo.rate('USD', 'EUR');
    expect(first).toBe(1.1);
    expect(second).toBe(1.1);
    expect(getRate).toHaveBeenCalledTimes(1);
  });

  it('returns null on quota denial', async () => {
    const getRate = vi.fn().mockResolvedValue(null);
    const repo = new FxRepository({ getRate } as any);
    const res = await repo.rate('USD', 'EUR');
    expect(res).toBeNull();
    expect(getRate).toHaveBeenCalled();
  });
});

File: mobile-app/test/fx_repository_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile_app/repositories/fx_repository.dart';
import 'package:smwa_services/services.dart';

class _FakeFxService extends FxService {
  int calls = 0;
  double? response;
  _FakeFxService({this.response});

  @override
  Future<double?> getRate(String from, String to) async {
    calls++;
    return response;
  }
}

void main() {
  group('FxRepository.rate', () {
    test('returns rate on success', () async {
      final svc = _FakeFxService(response: 1.1);
      final repo = FxRepository(service: svc);
      final rate = await repo.rate('USD', 'EUR');
      expect(rate, 1.1);
      expect(svc.calls, 1);
    });

    test('uses cache on repeat', () async {
      final svc = _FakeFxService(response: 1.2);
      final repo = FxRepository(service: svc);
      final first = await repo.rate('USD', 'EUR');
      final second = await repo.rate('USD', 'EUR');
      expect(first, same(second));
      expect(svc.calls, 1);
    });

    test('returns null on failure', () async {
      final svc = _FakeFxService(response: null);
      final repo = FxRepository(service: svc);
      final rate = await repo.rate('USD', 'EUR');
      expect(rate, isNull);
    });
  });
}
